File: js/BeginnersSelection/10_Traveling.js
// https://atcoder.jp/contests/abs/tasks/arc089_a

Object.prototype.let = function(f) {
  return f(this.valueOf());
}

require('fs')
  .readFileSync('/dev/stdin', 'utf8')
  .trim()
  .split('\n')
  .slice(1)
  .map(x => x.split(' ').map(x2 => +x2))
  .reduce((prev, next) =>
    prev
      && [prev, next].let(([[t1, x1, y1], [t2, x2, y2]]) => (
        (t2 & 1) === ((x2 + y2) & 1)
          && ((t2 - t1) >= (Math.abs(x2 - x1) + Math.abs(y2 - y1)))
      ))
      && next,
    [0, 0, 0]
  )
  .let(x => x ? 'Yes' : 'No')
  .let(console.log);

// https://atcoder.jp/contests/abs/submissions/21477949

File: rb/BeginnersSelection/10_Traveling.rb
# https://atcoder.jp/contests/abs/tasks/arc089_a

class Object
  def let(f)
    f.call self
  end
end

gets
  .to_i
  .let(->(n) { (1..n) })
  .map { |_| gets.split(" ").map(&:to_i) }
  .reduce([0, 0, 0]) { |p, n|
    p &&
      [p, n].let(->(((t1, x1, y1), (t2, x2, y2))) {
        (t2 & 1) === ((x2 + y2) & 1) &&
          ((t2 - t1) >= ((x2 - x1).abs + (y2 - y1).abs))
      }) &&
      n
  }
  .let(->(x) { x ? "Yes" : "No" })
  .let(method(:puts))

# https://atcoder.jp/contests/abs/submissions/21525732
